### PR TITLE
Improve test coverage

### PR DESCRIPTION
**Pull request by sparlane**
## Description

Improve the unit test coverage, and add end to end tests

## Summary by Sourcery

Expand automated testing and quality checks, adding end‑to‑end coverage and code coverage reporting around the existing server and client components.

Enhancements:
- Add generated certificate variants for alternative and expired client certs to support SSL failure scenarios in tests and expose a minimal Postgres schema for the server in the repo to support E2E runs.

Build:
- Wire configurable coverage flags through configure.ac and src/Makefile so coverage builds can be enabled cleanly, and add an opt-in `make e2e` target to run the end-to-end pytest suite.

CI:
- Replace the single C/C++ CI job with a multi-stage workflow that builds and runs tests, executes unit tests under valgrind, collects and uploads coverage to Codecov (including an HTML artifact), runs static analysis (cppcheck and clang-tidy), and runs E2E tests in CI.

Documentation:
- Document the new E2E test harness, its prerequisites, and usage, and add CI and Codecov badges to the main README.

Tests:
- Broaden C++ unit test coverage with new suites for logging, message queueing, reconnect behaviour, SSL failures, malformed and endianness/precision edge cases, and alignment regressions, plus shared test helpers.
- Stabilise existing connection and SSL tests by replacing sleep-based waits with a polling helper and adding SIGPIPE handling for the test binary.
- Introduce a Python-based E2E harness that drives the built server and fake client against a real Postgres/PostGIS instance via Docker, validating DB persistence, RTT, command delivery, server list handling, TLS cert rejection, and client reconnect semantics.

### CODE REVIEW

**Comment by sparlane** [comment]
@sourcery-ai review

**Comment by sparlane** [comment]
@sourcery-ai review

**Comment by sparlane** [comment]
@sourcery-ai review

File: .github/workflows/c-cpp.yml [hunk 1]
```diff
@@ -7,21 +7,213 @@ on:
...
     runs-on: ubuntu-latest
-
     steps:
-    - uses: actions/checkout@v6
-    - name: Install Deps
-      run: sudo apt update && sudo apt install -y libjsoncpp-dev libecpg-dev catch libgnutls28-dev gnutls-bin
+    - uses: actions/checkout@v4
+
+    - name: Install deps
+      run: |
```
Comment (SourceryAI) [inline .github/workflows/c-cpp.yml:17]: **issue (bug_risk):** The distcheck venv setup may fail on runners without python3-venv installed.

To avoid `make distcheck` failing on runners missing venv support, please add an explicit install step for `python3-venv` (and `python3-pip` if needed) in this job, similar to the `e2e` job. This will make the workflow more robust and consistent across environments.

File: src/Makefile.am [hunk 2]
```diff
@@ -8,6 +8,7 @@ AM_CPPFLAGS =
 AM_CFLAGS += $(CODE_COVERAGE_CFLAGS)
 AM_CXXFLAGS += $(CODE_COVERAGE_CXXFLAGS)
 AM_CPPFLAGS += $(CODE_COVERAGE_CPPFLAGS)
+AM_LDFLAGS = $(CODE_COVERAGE_LDFLAGS)
```
Comment (SourceryAI) [inline src/Makefile.am:11]: **suggestion (bug_risk):** Using `=` for AM_LDFLAGS can clobber existing linker flags; `+=` is likely safer.

Other coverage flags are appended (e.g. `AM_CFLAGS += ...`, `AM_CXXFLAGS += ...`), but here `AM_LDFLAGS` is overwritten with `=`. If `AM_LDFLAGS` is defined elsewhere, those flags will be lost when coverage is enabled. Using `AM_LDFLAGS += $(CODE_COVERAGE_LDFLAGS)` avoids clobbering existing linker flags.

File: tests/reconnect_test.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+     * these should produce an outbound connect. We can't observe that
+     * directly here, but we can observe wall-clock cost — 100 calls in
+     * <100ms is well under one per ms, which is only possible if each
+     * call is a no-op (actual connect() attempts to a closed port take
+     * a kernel round-trip each). */
+    auto start = std::chrono::steady_clock::now();
+    for (int i = 0; i < 100; ++i) { client->attemptReconnect(); }
+    auto elapsed = std::chrono::steady_clock::now() - start;
+    REQUIRE(std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count() < 500);
+}
```
Comment (SourceryAI) [inline tests/reconnect_test.cpp:119]: **suggestion (testing):** Timing-based assertion on `attemptReconnect` throttling may be brittle across environments

This test relies on 100 calls finishing in <500 ms to infer throttling behavior. On slower or contended CI machines that threshold may be exceeded even when throttling works, making the test flaky. Consider (a) relaxing the bound substantially, (b) treating it as a coarse regression check with a higher timeout (seconds) and logging, or (c) exposing the throttling logic to a controllable clock so it can be tested deterministically without wall-clock timing assumptions.

```suggestion
    /* Rapid-fire attemptReconnect: within the 1s retry window none of
     * these should produce an outbound connect. We can't observe that
     * directly here, but we can observe wall-clock cost as a coarse
     * regression check — 100 calls completing within a generous multi-
     * second bound indicates that attemptReconnect() remains a cheap
     * no-op when throttled (actual connect() attempts to a closed port
     * take a kernel round-trip each and would be substantially slower).
     *
     * NOTE: The timeout is intentionally loose to avoid flakiness on
     * slow or contended CI hosts; this is not a fine-grained perf test. */
    auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < 100; ++i) {
        client->attemptReconnect();
    }
    auto elapsed = std::chrono::steady_clock::now() - start;
    auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count();

    // Coarse regression bound: on healthy throttling this should typically be
    // orders of magnitude lower, but we allow several seconds for slow CI.
    REQUIRE(elapsed_ms < 5000);
}
```

File: e2e/conftest.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+        "assets_assetsearchprogress",
+        "assets_assetcommand",
+        "config_assetconfig",
+        "config_smmconfig",
+        "config_serverconfig",
+        "assets_asset",
+    ]
+    with db_conn.cursor() as cur:
+        # sourcery skip: sqlalchemy-execute-raw-query
+        cur.execute("TRUNCATE " + ", ".join(tables) + " RESTART IDENTITY CASCADE")
```
Comment (SourceryAI) [inline e2e/conftest.py:219]: **security (python.sqlalchemy.security.sqlalchemy-execute-raw-query):** Avoiding SQL string concatenation: untrusted input concatenated with raw SQL query can result in SQL Injection. In order to execute raw query safely, prepared statement should be used. SQLAlchemy provides TextualSQL to easily used prepared statement with named parameters. For complex SQL composition, use SQL Expression Language or Schema Definition Language. In most cases, SQLAlchemy ORM will be a better option.

*Source: opengrep*

File: e2e/conftest.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+                 "ca.tmpl", "client.tmpl", "localhost.tmpl"):
+        src = CERT_SCRIPTS / name
+        if src.exists():
+            shutil.copy(src, target / name)
+    for script in ("generate-ca.sh", "generate-server.sh", "generate-client.sh"):
+        (target / script).chmod(0o755)
+
+    # sourcery skip: dangerous-subprocess-use-audit
+    def run(cmd: List[str]) -> None:
+        subprocess.check_call(cmd, cwd=str(target))
```
Comment (SourceryAI) [inline e2e/conftest.py:244]: **security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'check_call' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*

File: e2e/conftest.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+    log_fp = log_path.open("wb")
+    # sourcery skip: dangerous-subprocess-use-audit
+    proc = subprocess.Popen(
+        [str(SERVER_BIN), str(config_path)],
+        cwd=str(REPO_ROOT),
+        env=env,
+        stdout=log_fp,
+        stderr=subprocess.STDOUT,
+    )
```
Comment (SourceryAI) [inline e2e/conftest.py:300]: **security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*

File: e2e/conftest.py [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+        env["LD_LIBRARY_PATH"] = str(REPO_ROOT / "src" / ".libs")
+        log_fp = log_path.open("wb")
+        # sourcery skip: dangerous-subprocess-use-audit
+        proc = subprocess.Popen(
+            [str(FAKE_CLIENT_BIN), str(config_path)],
+            cwd=str(REPO_ROOT),
+            env=env,
+            stdout=log_fp,
+            stderr=subprocess.STDOUT,
+        )
```
Comment (SourceryAI) [inline e2e/conftest.py:376]: **security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*

File: e2e/test_server_restart.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    env["PGPASSWORD"] = str(migrated_db["password"])
+
+    def spawn_server(log_name: str) -> tuple[subprocess.Popen, IO[bytes]]:
+        log_fp = (tmp_path / log_name).open("wb")
+        # sourcery skip: dangerous-subprocess-use-audit
+        p = subprocess.Popen(
+            [str(SERVER_BIN), str(config_path)],
+            cwd=str(REPO_ROOT), env=env,
+            stdout=log_fp, stderr=subprocess.STDOUT,
+        )
```
Comment (SourceryAI) [inline e2e/test_server_restart.py:69]: **security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*

File: e2e/test_server_restart.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        SERVER_PORT=str(port),
+    )
+    client_log = tmp_path / "client-test1.log"
+    client_fp = client_log.open("wb")
+    # sourcery skip: dangerous-subprocess-use-audit
+    client = subprocess.Popen(
+        [str(FAKE_CLIENT_BIN), str(client_cfg)],
+        cwd=str(REPO_ROOT), env=env,
+        stdout=client_fp, stderr=subprocess.STDOUT,
+    )
```
Comment (SourceryAI) [inline e2e/test_server_restart.py:108]: **security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 6 security issues, 3 other issues, and left some high level feedback:

**Security issues**:
- Avoiding SQL string concatenation: untrusted input concatenated with raw SQL query can result in SQL Injection. In order to execute raw query safely, prepared statement should be used. SQLAlchemy provides TextualSQL to easily used prepared statement with named parameters. For complex SQL composition, use SQL Expression Language or Schema Definition Language. In most cases, SQLAlchemy ORM will be a better option. ([link](https://github.com/canterbury-air-patrol/flight-safety-system/pull/44/files#diff-185939c99e317a14049225794c1f1df846a86efd9f849ec6095f17938991102aR219))
- Detected subprocess function 'check_call' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'. ([link](https://github.com/canterbury-air-patrol/flight-safety-system/pull/44/files#diff-185939c99e317a14049225794c1f1df846a86efd9f849ec6095f17938991102aR244))
- Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'. ([link](https://github.com/canterbury-air-patrol/flight-safety-system/pull/44/files#diff-185939c99e317a14049225794c1f1df846a86efd9f849ec6095f17938991102aR294-300))
- Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'. ([link](https://github.com/canterbury-air-patrol/flight-safety-system/pull/44/files#diff-185939c99e317a14049225794c1f1df846a86efd9f849ec6095f17938991102aR370-376))
- Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'. ([link](https://github.com/canterbury-air-patrol/flight-safety-system/pull/44/files#diff-957cfaaf4fe619dbb0cb5e3cbf10ee492f7992bab02c3815d26523d36edad3e1R65-69))
- Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'. ([link](https://github.com/canterbury-air-patrol/flight-safety-system/pull/44/files#diff-957cfaaf4fe619dbb0cb5e3cbf10ee492f7992bab02c3815d26523d36edad3e1R104-108))

**General comments**:

- Several C++ tests still hard‑code fixed TCP ports (e.g. 20502–20513) even though `fss_test::pick_port()` exists; switching those tests to use an ephemeral port helper will reduce flakiness from port conflicts when running tests in parallel or on busy CI hosts.
- In `src/Makefile.am` you assign `AM_LDFLAGS = $(CODE_COVERAGE_LDFLAGS)`, which overwrites any existing linker flags; consider using `+=` so coverage builds augment rather than replace other project-wide LDFLAGS.
- Some e2e tests rely on long `time.sleep(...)` windows (e.g. 12–16s) to wait for DB side effects; where possible, replacing these with polling helpers similar to `wait_for_row` will make the suite faster and less timing-sensitive.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- Several C++ tests still hard‑code fixed TCP ports (e.g. 20502–20513) even though `fss_test::pick_port()` exists; switching those tests to use an ephemeral port helper will reduce flakiness from port conflicts when running tests in parallel or on busy CI hosts.
- In `src/Makefile.am` you assign `AM_LDFLAGS = $(CODE_COVERAGE_LDFLAGS)`, which overwrites any existing linker flags; consider using `+=` so coverage builds augment rather than replace other project-wide LDFLAGS.
- Some e2e tests rely on long `time.sleep(...)` windows (e.g. 12–16s) to wait for DB side effects; where possible, replacing these with polling helpers similar to `wait_for_row` will make the suite faster and less timing-sensitive.

## Individual Comments

### Comment 1
<location path=".github/workflows/c-cpp.yml" line_range="34-17" />
<code_context>
     - name: make check
       run: make check
+
+    - name: prepare venv for distcheck
+      run: |
+        python3 -m venv .venv
+        source .venv/bin/activate
</code_context>
<issue_to_address>
**issue (bug_risk):** The distcheck venv setup may fail on runners without python3-venv installed.

To avoid `make distcheck` failing on runners missing venv support, please add an explicit install step for `python3-venv` (and `python3-pip` if needed) in this job, similar to the `e2e` job. This will make the workflow more robust and consistent across environments.
</issue_to_address>

### Comment 2
<location path="src/Makefile.am" line_range="11" />
<code_context>
 AM_CFLAGS += $(CODE_COVERAGE_CFLAGS)
 AM_CXXFLAGS += $(CODE_COVERAGE_CXXFLAGS)
 AM_CPPFLAGS += $(CODE_COVERAGE_CPPFLAGS)
+AM_LDFLAGS = $(CODE_COVERAGE_LDFLAGS)

 bin_PROGRAMS =
</code_context>
<issue_to_address>
**suggestion (bug_risk):** Using `=` for AM_LDFLAGS can clobber existing linker flags; `+=` is likely safer.

Other coverage flags are appended (e.g. `AM_CFLAGS += ...`, `AM_CXXFLAGS += ...`), but here `AM_LDFLAGS` is overwritten with `=`. If `AM_LDFLAGS` is defined elsewhere, those flags will be lost when coverage is enabled. Using `AM_LDFLAGS += $(CODE_COVERAGE_LDFLAGS)` avoids clobbering existing linker flags.
</issue_to_address>

### Comment 3
<location path="tests/reconnect_test.cpp" line_range="109-119" />
<code_context>
+     * <100ms is well under one per ms, which is only possible if each
+     * call is a no-op (actual connect() attempts to a closed port take
+     * a kernel round-trip each). */
+    auto start = std::chrono::steady_clock::now();
+    for (int i = 0; i < 100; ++i) { client->attemptReconnect(); }
+    auto elapsed = std::chrono::steady_clock::now() - start;
+    REQUIRE(std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count() < 500);
+}
+
</code_context>
<issue_to_address>
**suggestion (testing):** Timing-based assertion on `attemptReconnect` throttling may be brittle across environments

This test relies on 100 calls finishing in <500 ms to infer throttling behavior. On slower or contended CI machines that threshold may be exceeded even when throttling works, making the test flaky. Consider (a) relaxing the bound substantially, (b) treating it as a coarse regression check with a higher timeout (seconds) and logging, or (c) exposing the throttling logic to a controllable clock so it can be tested deterministically without wall-clock timing assumptions.

```suggestion
    /* Rapid-fire attemptReconnect: within the 1s retry window none of
     * these should produce an outbound connect. We can't observe that
     * directly here, but we can observe wall-clock cost as a coarse
     * regression check — 100 calls completing within a generous multi-
     * second bound indicates that attemptReconnect() remains a cheap
     * no-op when throttled (actual connect() attempts to a closed port
     * take a kernel round-trip each and would be substantially slower).
     *
     * NOTE: The timeout is intentionally loose to avoid flakiness on
     * slow or contended CI hosts; this is not a fine-grained perf test. */
    auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < 100; ++i) {
        client->attemptReconnect();
    }
    auto elapsed = std::chrono::steady_clock::now() - start;
    auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count();

    // Coarse regression bound: on healthy throttling this should typically be
    // orders of magnitude lower, but we allow several seconds for slow CI.
    REQUIRE(elapsed_ms < 5000);
}
```
</issue_to_address>

### Comment 4
<location path="e2e/conftest.py" line_range="219" />
<code_context>
        cur.execute("TRUNCATE " + ", ".join(tables) + " RESTART IDENTITY CASCADE")
</code_context>
<issue_to_address>
**security (python.sqlalchemy.security.sqlalchemy-execute-raw-query):** Avoiding SQL string concatenation: untrusted input concatenated with raw SQL query can result in SQL Injection. In order to execute raw query safely, prepared statement should be used. SQLAlchemy provides TextualSQL to easily used prepared statement with named parameters. For complex SQL composition, use SQL Expression Language or Schema Definition Language. In most cases, SQLAlchemy ORM will be a better option.

*Source: opengrep*
</issue_to_address>

### Comment 5
<location path="e2e/conftest.py" line_range="244" />
<code_context>
        subprocess.check_call(cmd, cwd=str(target))
</code_context>
<issue_to_address>
**security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'check_call' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*
</issue_to_address>

### Comment 6
<location path="e2e/conftest.py" line_range="294-300" />
<code_context>
    proc = subprocess.Popen(
        [str(SERVER_BIN), str(config_path)],
        cwd=str(REPO_ROOT),
        env=env,
        stdout=log_fp,
        stderr=subprocess.STDOUT,
    )
</code_context>
<issue_to_address>
**security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*
</issue_to_address>

### Comment 7
<location path="e2e/conftest.py" line_range="370-376" />
<code_context>
        proc = subprocess.Popen(
            [str(FAKE_CLIENT_BIN), str(config_path)],
            cwd=str(REPO_ROOT),
            env=env,
            stdout=log_fp,
            stderr=subprocess.STDOUT,
        )
</code_context>
<issue_to_address>
**security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*
</issue_to_address>

### Comment 8
<location path="e2e/test_server_restart.py" line_range="65-69" />
<code_context>
        p = subprocess.Popen(
            [str(SERVER_BIN), str(config_path)],
            cwd=str(REPO_ROOT), env=env,
            stdout=log_fp, stderr=subprocess.STDOUT,
        )
</code_context>
<issue_to_address>
**security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*
</issue_to_address>

### Comment 9
<location path="e2e/test_server_restart.py" line_range="104-108" />
<code_context>
    client = subprocess.Popen(
        [str(FAKE_CLIENT_BIN), str(client_cfg)],
        cwd=str(REPO_ROOT), env=env,
        stdout=client_fp, stderr=subprocess.STDOUT,
    )
</code_context>
<issue_to_address>
**security (python.lang.security.audit.dangerous-subprocess-use-audit):** Detected subprocess function 'Popen' without a static string. If this data can be controlled by a malicious actor, it may be an instance of command injection. Audit the use of this call to ensure it is not controllable by an external resource. You may consider using 'shlex.escape()'.

*Source: opengrep*
</issue_to_address>
~~~

</details>

***

Hi @sparlane! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Review by Shanaforth — APPROVED** [review]
